### ISSUE TITLE
Cant find rt.jar file error with JDK 10

### ISSUE THREAD

**Issue by adnan904** [eclipse-cognicrypt/CogniCrypt#202]
-    Ubuntu 18.04 LTS
-    Eclipse Photon 4.8.0
-    Oracle JDK 10

The static analyzer gives the "cannot find rt.jar" error  with JDK-10. The problem doesn't exist with JDK-8.
![screenshot from 2018-07-11 17-20-45](https://user-images.githubusercontent.com/34026698/42582225-e384796c-852e-11e8-909b-fda632d49065.png)

**Comment by johspaeth** [comment]
@adnan904, thanks for reporting this issue.

Does soot yet at all work with JDK 10, I assume the structure changed heavily and rt.jar does not exist anymore?
We use the [prepend-class path option](https://github.com/CROSSINGTUD/CogniCrypt/blob/511e3f021c5fc2925ebfb42ad2eea18956fe63e6/plugins/de.cognicrypt.staticanalyzer/src/de/cognicrypt/staticanalyzer/sootbridge/SootRunner.java#L120).

**Comment by kruegers** [comment]
I had a chat with @anddann about this issue. To resolve it, he suggests to switch to the [Java 9 soot branch](https://github.com/Sable/soot/tree/java9) as Java 9 got rid of the rt.jar, not Java 10. The readme in that branch describes how that would work. 
He also pointed out tho that even that soot version is currently using a byte code parser that is Java 9 but not (necessarily) Java 10 compatible. Hence, it might break when reading Java 10 code.

In any case, I think, we should indeed switch to that soot version rather soon because Java 10 is already out, Java 9 is no longer supported, and support for Java 8 is being discontinued by January 2019. What do you think, @johspaeth?

**Comment by johspaeth** [comment]
@mbenz @anddann is there a readily compiled soot SNAPSHOT version of the Java 9 branch on nexus that we can link CogniCrypt against it?

**Comment by kruegers** [comment]
Linking the right Manuel: @Mbenz89. Can you help, please?

**Comment by anddann** [comment]
Simple problem, we currently depend on asm-debug-all 6.0_BETA (the asm bytecode parser containing additional debug symbols and classes) to transform bytecode to jimple.
However, there exists currently no release for asm-all on maven central > 6.0 that contains all asm classes.
Thus,the asm frontend will throw an exception if soot is executed on anything higher than Java 9.

Nevertheless, what would work is putting the rt.jar from an older java version on the classpath.
That would be the way to go:
In general, how 'correct/reasonable' is an analysis of a project that has been compiled to java 8, when analysing it with JDK classes from Java 9,10, or 11?
That are NOT the classes the project was build for...

**Comment by johspaeth** [comment]
Thanks for the quick response. Do you think we should simply ship CogniCrypt with a rt.jar that it automatically links to?

**Comment by anddann** [comment]
Would be an options. Doop does something similar https://bitbucket.org/yanniss/doop-benchmarks
However, one could check on which version cognicrypt is executed and then download the lib on demand and cconfiguring soots arguments.

I guess shipping CogniCrypt with an rt.jar may lead to legal issues. I don't know but I would try to ship/devliver propertiy 3rd party code...

**Comment by anddann** [comment]
Just for completeness; looks like the classes are on maven-central but have been moved to a different package. The package name is now `asm-tree` for version 7.0.
Thus, one could try out to migrate the java 9 branch to this version and look what happens.

IMHO: Should work, but newer java Features are not supported.

@johspaeth @kruegers 
Do you have an SHK working on this? I could prepare something, but then the SHK needs to do the testing and writing JUnit Test cases?
Or someone else?

**Comment by kruegers** [comment]
Hm, @adnan904 would make sense since he experienced the issue first-hand already, but hes got other priorities for the time being. @johspaeth Could Rakshit take over this issue once Andi is done?

**Comment by johspaeth** [comment]
Yes, it makes sense for Rakshit to work on this. As of now, I would rather stick to the main branch of Soot and go with the solution @anddann described first, i.e., to download the appropriate jar.

**Comment by kruegers** [comment]
How come? I feel like going with the Java 9 branch is the more future-proof solution. Should that branch eventually be merged into the Soot's master branch, all we'd need to do is to point to a different branch not re-implement another solution. 
I'm not even saying I'm opposed to downloading the jar, but I wonder why you prefer it over the other solution.

**Comment by johspaeth** [comment]
@kruegers Soot propagates through all projects of CogniCrypt:
[CryptoAnalysis](https://github.com/CROSSINGTUD/CryptoAnalysis)
[boomerangPDS and idealPDS](https://github.com/CROSSINGTUD/WPDS) and must be linked to in all the dependencies. 
Also, having a completely different Java version on the Soot classpath can have side effects for the analysis (I observed some when switching form Java 1.7 to 1.8 where they reimplemented the HashMap and suddenly my test case time execution did massive change).

Just don't want to make it too much of work for either @anddann or Rakshit. But if you,
@anddann can make a precompiled version of Soot of that branch available on the soot nightly build server such that we only have to update our pom.xml, we could give it a try.

**Comment by mbenz89** [comment]
The repository is browsable. This is the last snapshot build of soot-j9 (7.10.18):"https://soot-build.cs.uni-paderborn.de/nexus/#browse/browse/components:soot-snapshot:36e3dec8de528c9bac5cfa50b607992c

@andann is j9 stable enough to merge it back into master? We are getting more and more request for it... 
Also, we should include some notes about the impact of Soot's execution environment on the analysis to the documentation.

Sry for hijacking this issue :)

**Comment by anddann** [comment]
> @AnDann is j9 stable enough to merge it back into master? We are getting more and more request for it...

IMHO: it is stable enough, but there are two minor things I would like to clarify with you before merging the java9 branch into master. I would like to know your opinion on 1-2 code locations I still consider ugly/hacky. The rest of the codebase is fine, IMHO.


@johspaeth 
For this experiment, I don't want to make a version directly availabe on the nightly build server, which requires merging directly into the soot java9 branch.

I propose the following process (feedback is welcome):
- I would fork the soot repository and work on the fork (to not break existing code)
- I adapt the soot java9branch, (in the fork) to work with the newest asm bytecode parser (max 1-2 h)
- Rakshit tries the fork on java 11 and writes a few junit test, testing if soot properly reads classes from the JDK 11
- Rakshit pushes the test into my fork
- if everything work, we could merge it back to the java 9 branch

! Keep in mind, that none of languages features >= Java 10 will be supported in jimple. We just adapt Soot to read classes contained in newer JDK version!

**Comment by mbenz89** [comment]
The repository is browsable. This is the last snapshot build of soot-j9
(7.10.18):"
https://soot-build.cs.uni-paderborn.de/nexus/#browse/browse/components:soot-snapshot:36e3dec8de528c9bac5cfa50b607992c

@andann is j9 stable enough to merge it back into master? We are getting
more and more request for it...
Also, we should include some notes about the impact of Soot's execution
environment on the analysis to the documentation.

Sry for hijacking this issue :)

Johannes Späth <notifications@github.com> schrieb am Fr., 19. Okt. 2018,
18:54:

**Comment by rakshitkr** [comment]
@johspaeth Please add me as a collaborator so that I can self-assign this issue.

**Comment by rakshitkr** [comment]
@anddann Where should I implement the test cases for soot?

**Comment by johspaeth** [comment]
Thanks @anddann and @rakshitkr. 
@anddann the process you suggested is ok for me.

I guess @rakshitkr shall implement a couple of minimal classes that need to load classes from the JDK of Java 9/Java 10 and 11 into Soot's Scene?

**Comment by anddann** [comment]
Yes, execatly.

> Thanks @anddann and @rakshitkr.
> @anddann the process you suggested is ok for me.
> 
> I guess @rakshitkr shall implement a couple of minimal classes that need to load classes from the JDK of Java 9/Java 10 and 11 into Soot's Scene?

Yes, exactly. That's all.
@rakshitkr if I've modified Soot, I let you know. However, I won't make it until friday or next week.

**Comment by rakshitkr** [comment]
@anddann can I proceed with my implementation?

**Comment by anddann** [comment]
Hi @rakshitkr 
I've pushed my changes to https://github.com/anddann/soot/tree/fix/newAsm
This branch should execute successfully on JDK9-12.
In the Readme.md you find all information how to execute the branch using the command line.
Could you please write some JUnit tests that test the branch when loading JDK9-12 classes.
If you have any problems or questions left do not hesitate to ask @kruegers

**Comment by rakshitkr** [comment]
@anddann Thank you very much for the update. Will work on this on Wednesday and communicate the results very soon.

**Comment by kruegers** [comment]
assigned @rakshitkr.

**Comment by johspaeth** [comment]
@rakshitkr Please finish this issue by resolving https://github.com/CROSSINGTUD/CryptoAnalysis/issues/143 first
@sriteja As a workaround please implement a version check and warn the user, if (s)he need to change the [-vm argument in the eclipse.ini](https://matsim.org/docs/devguide/eclipse/jdk) to use JDK/JRE 8 or lower.

**Comment by kruegers** [comment]
@rakshitkr & @sritejakv After merging #311, this issue is resolved, right?

**Comment by rakshitkr** [comment]
@kruegers yes!

**Comment by rakshitkr** [comment]
@kruegers Kindly close this issue if there is nothing else that needs to be implemented.

**Comment by kruegers** [comment]
@rakshitkr It is closed already. Since September 10th.
